src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala

```diff
@@ -0,0 +1,181 @@
+package mesosphere.marathon.core.task.termination.impl
+
+import akka.actor.{ Actor, ActorLogging, Cancellable, Props }
+import mesosphere.marathon.MarathonSchedulerDriverHolder
+import mesosphere.marathon.core.base.Clock
+import mesosphere.marathon.core.task.termination.TaskKillConfig
+import mesosphere.marathon.state.Timestamp
+import mesosphere.marathon.core.task.{ Task, TaskStateOp }
+import mesosphere.marathon.core.task.tracker.{ TaskStateOpProcessor, TaskTracker }
+import mesosphere.marathon.core.event.MesosStatusUpdateEvent
+
+import scala.collection.mutable
+import scala.concurrent.Promise
+
+/**
+  * An actor that handles killing tasks in chunks and depending on the task state.
+  * Lost tasks will simply be expunged from state, while active tasks will be killed
+  * via the scheduler driver. There is be a maximum number of kills in flight, and
+  * the service will only issue more kills when tasks are reported terminal.
+  */
+private[impl] class TaskKillServiceActor(
+    taskTracker: TaskTracker,
+    driverHolder: MarathonSchedulerDriverHolder,
+    stateOpProcessor: TaskStateOpProcessor,
+    config: TaskKillConfig,
+    clock: Clock) extends Actor with ActorLogging {
+  import TaskKillServiceActor._
+  import context.dispatcher
+
+  val tasksToKill: mutable.HashMap[Task.Id, Option[Task]] = mutable.HashMap.empty
+  val inFlight: mutable.HashMap[Task.Id, TaskToKill] = mutable.HashMap.empty
+  var retryTimer: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[MesosStatusUpdateEvent])
+  }
+
+  //scalastyle:off cyclomatic.complexity
+  override def receive: Receive = {
+    case KillTaskById(taskId) =>
+      killTaskById(taskId)
+
+    case KillTask(task) =>
+      killTask(task)
+
+    case KillUnknownTaskById(taskId) =>
+      killUnknownTaskById(taskId)
+
+    case KillTasks(tasks, promise) =>
+      killTasks(tasks, promise)
+
+    case Terminal(event) if inFlight.contains(event.taskId) || tasksToKill.contains(event.taskId) =>
+      handleTerminal(event.taskId)
+
+    case Retry =>
+      retry()
+
+    case unhandled: Request =>
+      log.warning("Received unhandled {}", unhandled)
+
+    case unexpected: Any =>
+      log.debug("Received unexpected {}", unexpected)
+  }
+  //scalastyle:on
+
+  def killTaskById(taskId: Task.Id): Unit = {
+    log.debug("Received KillTaskById({})", taskId)
+    import context.dispatcher
+    taskTracker.task(taskId).map {
```

Comment:
Good question. My intention was to have a 1:1 mapping of `Message -> Action`, that's why I chose to use the existing Messages in this case as well. Afaik we have both ways of doing it in our code base, maybe it's time for a style direction and picking one best practice/convention. Preferences, anybody?

That said, I realize the only place where we call `killTaskbyId` is via the TaskReplaceActor, which performs a lossy mapping of `Iterable[Task]` to a Set[Task.Id] (although it has all the tasks in scope all the way). I figure by changing that to kill [Task]'s explicitly, we can get rid of `killTaskbyId` completely. Which makes sense imo, because requesting a task to be killed without having its state at hand seems like killing it blindfold ..